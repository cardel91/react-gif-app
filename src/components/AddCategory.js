import React from 'react'
import { useState } from 'react'
import PropTypes from 'prop-types';

export const AddCategory = ({setCategories}) => {



    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (event) => {
        setInputValue(event.target.value);
    }

    const handleSubmit = (e)=>{
        e.preventDefault();

        // validacion de inputvalue
        if(inputValue.trim().length >2){
            // llamar con el callback del estado anterior
            setCategories(c=>[inputValue, ...c ]);
            setInputValue('');
        }
        // this.props.setCategories(e.target.value);
    }

    return (
        <form onSubmit={handleSubmit}>
            <input 
                type="text" 
                value = {inputValue}
                onChange = {handleInputChange}
            />
        </form>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
