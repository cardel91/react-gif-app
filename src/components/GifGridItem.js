import React from 'react'
import PropTypes from 'prop-types';

// var count = 0;
// export const GifGridItem = (props) => {
    // count++;
    // console.log(count);
export const GifGridItem = ({id, title, url}) => {
    // console.log(id, title, url);
    return (
        <div className='card animate__animated animate__bounce'>
            <img src={url} alt={title}/>
            <p>{title}</p>
        </div>
    )
}

GifGridItem.propTypes = {
    id: PropTypes.string,
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
}