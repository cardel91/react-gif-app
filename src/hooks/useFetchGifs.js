import { useEffect, useState } from "react"
import { getGifs } from "../helpers/GetGifs";


// 
export const useFetchGifs = (category) => {
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    // // recibe una funcion y una lista de dependencias
    // // si la lista es vacia useEffect solo se lanza una vez
    // // similar a componentDidMount() de las clases
    // useEffect(()=>{
    //     // getGifs(category).then(imgs=>setImages(imgs));

    //     // se puede hacer referencia al metodo sin parentesis
    //     // si solo recibe el parametro que devuelve el callback
    //     getGifs(category).then(setImages);
    // /**
    //  * Para este error
    //  *   Line 18:8:  React Hook useEffect has a missing dependency: 'category'. Either include it or remove the dependency array  react-hooks/exhaustive-deps
    //  * se debe agregar la categoria, ya que si la categoria cambia
    //  * deberÃ¡ ejecutarse nuevamente
    //  */
    // }, [category]);

    // useEffect no puede ser async
    useEffect(()=>{
        getGifs(category).then(imgs => {
            // setTimeout(() => {
                console.log(imgs);
                setState({
                    data: imgs,
                    loading: false
                });
            // }, 2300);
        });
       
    }, [category]);

    return state;
}