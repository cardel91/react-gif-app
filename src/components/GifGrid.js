import React/*,{ useState, useEffect }*/ from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs'
// import { getGifs } from '../helpers/GetGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({category}) => {

    // const [images, setImages] = useState([]);


    
    // enviar category como prop
    // alias: images para propiedad data
    const {data: images, loading} = useFetchGifs(category);
    // console.log(state);
    return (

        <>
            <h3 className="animate__animated animate__fadeIn">{category}</h3>

            { loading && <p className="animate__animated animate__flash">'Cargando...'</p>}
            <div className='card-grid'>
                {/* <ol> */}
                    {
                        images.map(img =>{
                        // desestructurar el objeto img
                        // images.map(({id, title, url}) =>{

                            return(
                                <GifGridItem key={img.id} 
                                // 
                                {...img}
                                />
                            );
                        })
                    }
                {/* </ol> */}
            </div>
        </>
    )
}
