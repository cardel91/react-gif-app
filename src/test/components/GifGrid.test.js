import { GifGrid } from "../../components/GifGrid";
import { shallow } from 'enzyme';
import { useFetchGifs } from "../../hooks/useFetchGifs";
import '@testing-library/jest-dom';

// hace un mock del hook useFetchGifs
jest.mock("../../hooks/useFetchGifs");

describe('Pruebas GifGrid', () => {
   

    const category = 'Ola k ase';
    
    test('debe hacer match con el snapshot', () => {
        // falsa implementacion
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow(<GifGrid category={category}/> )

        expect(wrapper).toMatchSnapshot();
    });

    test('debe mostrar items cuando se carga imagenes con useFetchGifs', () => {
        
        const gifs = [
            {
                id: 'ABC',
                title: 'Ola k ase',
                url: 'https://localhost.gif'
            },
            {
                id: '245',
                title: 'Ola k ase',
                url: 'https://localhost.gif'
            }
        ];

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        const wrapper = shallow(<GifGrid category={category}/> )

        // expect(wrapper).toMatchSnapshot();

        expect(wrapper.find('p').exists()).toBe(false);

        expect(wrapper.find('GifGridItem').length).toBe(gifs.length);

    });
    
    
});
