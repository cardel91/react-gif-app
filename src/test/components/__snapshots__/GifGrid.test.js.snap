// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas GifGrid debe hacer match con el snapshot 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Ola k ase
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    'Cargando...'
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
